File: projects/heart-linker-app/ios/Meishi/Models/SnsAccount.swift
import Foundation

// MARK: - SnsAccount

/// SNSアカウント1件。3テーブル (profile_sns / profile_draft_sns / contact_sns) を
/// table / parentColumn を切り替えることで同一クラスで扱う ActiveRecord パターン。
final class SnsAccount {
    var id: Int64 = 0
    var parentId: Int64
    var snsType: SnsType
    var value: String
    var serviceName: String   // type == .other のときにユーザーが入力した自由なサービス名
    var accountUrl: String?
    var accountId: String?
    var sortOrder: Int

    private let table: String
    private let parentColumn: String

    init(id: Int64 = 0,
         parentId: Int64,
         snsType: SnsType,
         value: String,
         serviceName: String = "",
         accountUrl: String? = nil,
         accountId: String? = nil,
         sortOrder: Int = 0,
         table: String,
         parentColumn: String) {
        self.id = id
        self.parentId = parentId
        self.snsType = snsType
        self.value = value
        self.serviceName = serviceName
        self.accountUrl = accountUrl
        self.accountId = accountId
        self.sortOrder = sortOrder
        self.table = table
        self.parentColumn = parentColumn
    }

    /// 表示名: OTHER のときはユーザー入力名、それ以外は固定名
    var displayLabel: String {
        snsType == .other && !serviceName.isEmpty ? serviceName : snsType.displayName
    }

    // MARK: Save / Delete

    @discardableResult
    func save() -> Int64 {
        let db = Database.shared
        let params: [SqlValue] = [
            .int(parentId), .text(snsType.rawValue), .text(value),
            .text(serviceName), .opt(accountUrl), .opt(accountId), .int(Int64(sortOrder))
        ]
        if id <= 0 {
            db.run("""
                INSERT INTO \(table) (\(parentColumn), type, value, service_name, account_url, account_id, sort_order)
                VALUES (?,?,?,?,?,?,?)
                """, params)
            id = db.lastInsertRowId()
        } else {
            db.run("""
                UPDATE \(table) SET \(parentColumn)=?, type=?, value=?, service_name=?, account_url=?, account_id=?, sort_order=?
                WHERE id=\(id)
                """, params)
        }
        return id
    }

    func delete() {
        guard id > 0 else { return }
        Database.shared.run("DELETE FROM \(table) WHERE id=?", [.int(id)])
    }

    // MARK: - Factory

    static func forProfile(profileId: Int64, snsType: SnsType, value: String, sortOrder: Int, serviceName: String = "") -> SnsAccount {
        SnsAccount(parentId: profileId, snsType: snsType, value: value, serviceName: serviceName, sortOrder: sortOrder,
                   table: "profile_sns", parentColumn: "profile_id")
    }
    static func forDraft(draftId: Int64, snsType: SnsType, value: String, sortOrder: Int, serviceName: String = "") -> SnsAccount {
        SnsAccount(parentId: draftId, snsType: snsType, value: value, serviceName: serviceName, sortOrder: sortOrder,
                   table: "profile_draft_sns", parentColumn: "draft_id")
    }
    static func forContact(contactId: Int64, snsType: SnsType, value: String, sortOrder: Int,
                           serviceName: String = "", accountUrl: String? = nil, accountId: String? = nil) -> SnsAccount {
        SnsAccount(parentId: contactId, snsType: snsType, value: value, serviceName: serviceName,
                   accountUrl: accountUrl, accountId: accountId, sortOrder: sortOrder,
                   table: "contact_sns", parentColumn: "contact_id")
    }

    // MARK: - Query helpers

    static func findAll(table: String, parentColumn: String, parentId: Int64) -> [SnsAccount] {
        var rows: [SnsAccount] = []
        Database.shared.query(
            "SELECT * FROM \(table) WHERE \(parentColumn)=? ORDER BY sort_order",
            [.int(parentId)]
        ) { row in
            rows.append(SnsAccount(
                id: row.int64(0),
                parentId: row.int64(1),
                snsType: SnsType(rawValue: row.text(3)) ?? .other,   // col3 = type
                value: row.text(4),                                    // col4 = value
                serviceName: row.text(2),                              // col2 = service_name
                accountUrl: row.textOrNil(5),
                accountId: row.textOrNil(6),
                sortOrder: row.int(7),
                table: table, parentColumn: parentColumn
            ))
        }
        return rows
    }

    static func deleteAll(table: String, parentColumn: String, parentId: Int64) {
        Database.shared.run("DELETE FROM \(table) WHERE \(parentColumn)=?", [.int(parentId)])
    }
}

File: projects/heart-linker-app/android/app/src/main/java/com/meishi/app/model/SnsAccount.kt
package com.meishi.app.model

import android.content.ContentValues
import android.database.Cursor
import com.meishi.app.db.ActiveRecord

/**
 * SNSアカウント1件分のActiveRecord。
 * table/parentColumnを切り替えることで profile_sns / profile_draft_sns / contact_sns
 * の3テーブルを同じクラスで扱う。
 *
 * value: 編集画面でユーザーが入力する生のテキスト(URLでもIDでも可)。
 * serviceName: typeがOTHERの場合にユーザーが入力するSNS名。
 * accountUrl/accountId: MessagePack交換時に value から推定して埋める(送受信用)。
 */
class SnsAccount(
    override var id: Long = 0,
    var parentId: Long = 0,
    var type: SnsType = SnsType.WEBSITE,
    var value: String = "",
    var serviceName: String = "",
    var accountUrl: String? = null,
    var accountId: String? = null,
    var sortOrder: Int = 0,
    private val table: String = TABLE_PROFILE,
    private val parentColumn: String = COLUMN_PROFILE_ID
) : ActiveRecord() {

    override val tableName: String get() = table

    override fun toContentValues(): ContentValues = ContentValues().apply {
        put(parentColumn, parentId)
        put("type", type.typeKey)
        put("value", value)
        put("service_name", serviceName)
        put("account_url", accountUrl)
        put("account_id", accountId)
        put("sort_order", sortOrder)
    }

    /** 表示名(OTHER種別ならユーザー入力名、それ以外は固定名) */
    fun displayLabel(): String = if (type == SnsType.OTHER && serviceName.isNotBlank()) serviceName else type.displayName

    companion object {
        const val TABLE_PROFILE = "profile_sns"
        const val TABLE_DRAFT = "profile_draft_sns"
        const val TABLE_CONTACT = "contact_sns"

        const val COLUMN_PROFILE_ID = "profile_id"
        const val COLUMN_DRAFT_ID = "draft_id"
        const val COLUMN_CONTACT_ID = "contact_id"

        private fun fromCursor(c: Cursor, table: String, parentColumn: String): SnsAccount {
            return SnsAccount(
                id = c.getLong(c.getColumnIndexOrThrow("id")),
                parentId = c.getLong(c.getColumnIndexOrThrow(parentColumn)),
                type = SnsType.fromKey(c.getString(c.getColumnIndexOrThrow("type"))),
                value = c.getString(c.getColumnIndexOrThrow("value")),
                serviceName = c.getString(c.getColumnIndexOrThrow("service_name")),
                accountUrl = c.getString(c.getColumnIndexOrThrow("account_url")),
                accountId = c.getString(c.getColumnIndexOrThrow("account_id")),
                sortOrder = c.getInt(c.getColumnIndexOrThrow("sort_order")),
                table = table,
                parentColumn = parentColumn
            )
        }

        fun findAllForProfile(profileId: Long) =
            findAllFor(TABLE_PROFILE, COLUMN_PROFILE_ID, profileId)

        fun findAllForDraft(draftId: Long) =
            findAllFor(TABLE_DRAFT, COLUMN_DRAFT_ID, draftId)

        fun findAllForContact(contactId: Long) =
            findAllFor(TABLE_CONTACT, COLUMN_CONTACT_ID, contactId)

        private fun findAllFor(table: String, parentColumn: String, parentId: Long): List<SnsAccount> {
            val list = mutableListOf<SnsAccount>()
            val cursor = rawQuery(
                "SELECT * FROM $table WHERE $parentColumn = ? ORDER BY sort_order ASC",
                arrayOf(parentId.toString())
            )
            cursor.use {
                while (it.moveToNext()) {
                    list.add(fromCursor(it, table, parentColumn))
                }
            }
            return list
        }

        fun deleteAllFor(table: String, parentColumn: String, parentId: Long) {
            writableDb().delete(table, "$parentColumn = ?", arrayOf(parentId.toString()))
        }

        fun newForProfile(profileId: Long, type: SnsType, value: String, sortOrder: Int, serviceName: String = "") =
            SnsAccount(parentId = profileId, type = type, value = value, serviceName = serviceName, sortOrder = sortOrder,
                table = TABLE_PROFILE, parentColumn = COLUMN_PROFILE_ID)

        fun newForDraft(draftId: Long, type: SnsType, value: String, sortOrder: Int, serviceName: String = "") =
            SnsAccount(parentId = draftId, type = type, value = value, serviceName = serviceName, sortOrder = sortOrder,
                table = TABLE_DRAFT, parentColumn = COLUMN_DRAFT_ID)

        fun newForContact(
            contactId: Long,
            type: SnsType,
            value: String,
            sortOrder: Int,
            serviceName: String = "",
            accountUrl: String? = null,
            accountId: String? = null
        ) = SnsAccount(
            parentId = contactId, type = type, value = value, serviceName = serviceName,
            accountUrl = accountUrl, accountId = accountId, sortOrder = sortOrder,
            table = TABLE_CONTACT, parentColumn = COLUMN_CONTACT_ID
        )
    }
}
